[Merged by Bors] - feat(data/finset/noncomm_prod): finite pi lemmas

nomeata opened the pull request:
including a few helpers

Co-authored-by: Junyan Xu <junyanxumath@gmail.com>

---

Extracted from https://github.com/leanprover-community/mathlib/pull/11744#discussion_r814107692

<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message 
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

- [x] depends on: #12473
- [x] depends on: #12474

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

nomeata committed:
feat(data/list/prod_monoid): finite pi lemmas

including a few helpers

nomeata commented on src/data/finset/noncomm_prod.lean:
```diff
@@ -217,6 +217,43 @@ by simp [noncomm_prod, insert_val_of_not_mem ha, multiset.noncomm_prod_cons']
     (λ x hx y hy, by rw [mem_singleton.mp hx, mem_singleton.mp hy]) = f a :=
 by simp [noncomm_prod, multiset.singleton_eq_cons]
 
+
```
```suggestion
```

nomeata committed:
Update src/data/finset/noncomm_prod.lean

eric-wieser commented on src/data/list/prod_monoid.lean:
```diff
@@ -26,6 +26,14 @@ begin
 end
 
+@[to_additive]
+lemma prod_eq_pow_of_forall_eq [monoid α] (l : list α) (m : α) (h : ∀ (x : α), x ∈ l → x = m) :
```
```suggestion
lemma prod_eq_of_forall_eq [monoid α] (l : list α) (m : α) (h : ∀ (x : α), x ∈ l → x = m) :
```
To match the `le` version below. Or change that one, since having pow in the name probably makes it easier to find both

nomeata replied:
Will change the `le` lemma

eric-wieser commented on src/algebra/group/pi.lean:
```diff
@@ -235,6 +236,21 @@ lemma pi.single_mul [Π i, mul_zero_class $ f i] (i : I) (x y : f i) :
 
+@[to_additive]
+lemma pi.mul_single_commute [Π i, mul_one_class $ f i] :
+  pairwise (λ i j, ∀ (x : f i) (y : f j), commute (mul_single i x) (mul_single j y)) :=
```
Might be nice to also add
```lean
lemma commute.mul_single {x y : f i} (h : commute x y) : commute (mul_single i x) (mul_single i y) :=
```
(for dot notation)

nomeata replied:
That'd be just `commute.map` with `momoid_hom.single`, right? Still with adding? But I guess yes, we have the other homomorphism properties of the unbundled `mul_single` as lemmas as well.

eric-wieser replied:
Ah good point, then that's not all that helpful after all. Maybe just mention that in the docstring of this one; "for when i and j are equal but x and y commute, `commute.map` can be used" or similar

eric-wieser commented on src/algebra/group/pi.lean:
```diff
@@ -235,6 +236,21 @@ lemma pi.single_mul [Π i, mul_zero_class $ f i] (i : I) (x y : f i) :
+end
+
+@[to_additive]
+lemma monoid_hom.single_commute [Π i, mul_one_class $ f i] :
```
I'd argue there's no point having this one, since `monoid_hom.single_apply` can fire first through `simp`. It's similar to how we tend not to state injectivity of `submonoid.subtype`, since that gets simplified away too

nomeata replied:
Hmm, the lack of a lemma for injectivity of subtype has been a regular pain for me in the nilpotency work, because it meant I couldn't easily instantiate lemmas directly.

nomeata replied:
But yes, in this case I don’t need this lemma and can easily use the other one. Removed.

eric-wieser commented on src/data/finset/noncomm_prod.lean:
```diff
@@ -217,6 +217,42 @@ by simp [noncomm_prod, insert_val_of_not_mem ha, multiset.noncomm_prod_cons']
+lemma noncomm_prod_map {α : Type*} {β : Type*} [monoid β]
+  (s : finset α) (f : α → β)
+  (comm : ∀ (x : α), x ∈ s → ∀ (y : α), y ∈ s → commute (f x) (f y))
+  {M : Type*} [monoid M] (g : β →* M) :
```
Can you generalize this in the same way that `commute.map` is generalized?

nomeata replied:
With generalize, you mean 
```patch
@@ -222,7 +222,7 @@ by simp [noncomm_prod, multiset.singleton_eq_cons]
 lemma noncomm_prod_map {α : Type*} {β : Type*} [monoid β]
   (s : finset α) (f : α → β)
   (comm : ∀ (x : α), x ∈ s → ∀ (y : α), y ∈ s → commute (f x) (f y))
-  {M : Type*} [monoid M] (g : β →* M) :
+  {M : Type*} [monoid M] {F : Type*} [monoid_hom_class F β M] (g : F) :
   g (s.noncomm_prod f comm) = s.noncomm_prod (λ i, g (f i))
   (λ x hx y hy, commute.map (comm x hx y hy) g)  :=
 begin
```

That would require me to generalize [monoid_hom.map_list_prod](https://leanprover-community.github.io/mathlib_docs/data/list/big_operators.html#monoid_hom.map_list_prod) as well, it seems. Is that what you have in mind?

nomeata replied:
Generalized lemmas as far as possible, but this looses dot-notation in a few cases, and in one case a lemma wouldn’t apply anymore if I don’t use `change` before to help lean a bit.

Is this how it should be?

Also, I don’t fully understand when to use `→*` and when `monoid_hom_class`? The docs say that one should always phrase lemmas in terms of the latter, but there are plenty of lemmas that don’t – is that a recent change that hasn’t been applied everywhere yet?

eric-wieser replied:
@Vierkantor can probably help here

nomeata replied:
The fallout is rather big. I'd suggest I go back to the version I had before, so that this PR doesn't conflate adding new lemmas with refactoring existing ones, and we leave an overhaul of existing API to use `monoid_hom_class` for another time. Would that be ok with you, @eric-wieser?

Vierkantor replied:
> Also, I don’t fully understand when to use `→*` and when `monoid_hom_class`? The docs say that one should always phrase lemmas in terms of the latter, but there are plenty of lemmas that don’t – is that a recent change that hasn’t been applied everywhere yet?

Exactly, this was changed only at the end of last year and we still need to update all the hundreds of usages.

I'd be happy to make a few separate refactoring PRs if you feel that would be too much for this one. It's definitely time to tackle `map_{list,multiset,finset,finsupp}_{prod,sum}`.

nomeata replied:
I'd like to keep this out of the critical path for my stuff, as it's rather tangential, thanks.

Is this then good to go as is?

nomeata replied:
Ah you approved, so someone can bors it now?

Vierkantor replied:
`map_{list,multiset,finset,finsupp}_prod`: #12354

eric-wieser replied:
Just to check: Does this mean the generalization is now trivial, or are there upstream lemmas that still need generalizing first?

eric-wieser replied:
Just tried: all the lemmas are in place now, so all you need to do is change the lemmas that you're adding in this PR.

nomeata committed:
Rename to prod_le_pow_of_forall_le

nomeata committed:
Docstring

nomeata committed:
Remove monoid_hom.single_commute

nomeata committed:
Generalize noncomm_prod_map and its dependencies

nomeata committed:
Line length

nomeata committed:
Fix fallout

nomeata committed:
Fix more fallout

nomeata committed:
More monoid_hom_class fallout

nomeata committed:
More monoid_hom_class fallout

nomeata committed:
More monoid_hom_class fallout

nomeata committed:
Undo changing `map_list_prod` to `monoid_hom_class`

Vierkantor reviewed (approved):
bors d=@eric-wieser

eric-wieser committed:
Merge remote-tracking branch 'origin/master' into joachim/finite_pi_ext

eric-wieser commented on src/data/finset/noncomm_prod.lean:
```diff
@@ -217,6 +217,42 @@ by simp [noncomm_prod, insert_val_of_not_mem ha, multiset.noncomm_prod_cons']
 
+@[to_additive]
+lemma noncomm_prod_map {α : Type*} {β : Type*} [monoid β]
+  (s : finset α) (f : α → β)
```
I think there is already an `α` and `β` of the right type floating around
```suggestion
lemma noncomm_prod_map (s : finset α) (f : α → β)
```
It would probably make sense to rename `M` to `γ` and put it right at the top of the file too.

eric-wieser commented on src/algebra/group/pi.lean:
```diff
@@ -235,6 +236,20 @@ lemma pi.single_mul [Π i, mul_zero_class $ f i] (i : I) (x y : f i) :
 
+/-- The injection into a pi group at different indices commutes.
+
+For injections of commmuting elements at the same index, see `commute.map`
```
```suggestion
For injections of commuting elements at the same index, see `commute.map`
```

eric-wieser commented on src/algebra/group/pi.lean:
```diff
@@ -235,6 +236,20 @@ lemma pi.single_mul [Π i, mul_zero_class $ f i] (i : I) (x y : f i) :
+  by_cases h2 : j = k, { subst h2, simp [hij], },
+  simp [h1,  h2],
+end
+
```
This feels like it might be a good lemma too, since you prove this below as part of a lemma statement:
```suggestion

/-- The injection into a pi group with the same values commutes. -/
@[to_additive "The injection into a pi additive group with the same values commutes."]
lemma pi.mul_single_apply_commute [Π i, mul_one_class $ f i] (x : Π i, f i) (i j : I) :
  commute (mul_single i (x i)) (mul_single j (x j)) :=
begin
  obtain rfl | hij := decidable.eq_or_ne i j,
  { refl },
  { exact pi.mul_single_commute _ _ hij _ _, },
end

```

nomeata replied:
Good idea

eric-wieser commented on src/algebra/group/pi.lean:
```diff
@@ -235,6 +236,20 @@ lemma pi.single_mul [Π i, mul_zero_class $ f i] (i : I) (x y : f i) :
+
+For injections of commmuting elements at the same index, see `commute.map`
+-/
+@[to_additive]
```
Can you `to_additive` the docstring too?

nomeata replied:
Is it just a matter of reversing the order, or do I have to repeat the docstring in the attribute? (Assuming the same text works for both)

I'll read up on it.

eric-wieser commented on src/data/finset/noncomm_prod.lean:
```diff
@@ -244,4 +280,39 @@ begin
+@[to_additive]
+lemma noncomm_prod_single {ι : Type*} [fintype ι] [decidable_eq ι]
+  {M : ι → Type*} [∀ i, monoid (M i)] (x : Π i, M i) :
+  finset.univ.noncomm_prod (λ i, monoid_hom.single M i (x i))
```
I think this should probably be stated about `mul_single`, since that's the simp-normal form?

nomeata replied:
Is that so? I feat it might become tedious having to deal with `mul_single` when you think of it as a group homomorphism, but that bundled information (if that’s the right term) gets thrown away.

But I’ll try.

nomeata replied:
I think I did the simp setup for `monoid_hom.single` wrong, but I think I have fixed it. With this fix, I could use `pi.mul_single` much more widely in #11744, so looks good.

eric-wieser commented on src/data/finset/noncomm_prod.lean:
```diff
@@ -244,4 +280,39 @@ begin
+  apply (finset.univ.noncomm_prod_map (λ i, monoid_hom.single M i (x i)) _
+    (pi.eval_monoid_hom _ i)).trans,
+  rw (finset.insert_erase (finset.mem_univ i)).symm,
+  rw finset.noncomm_prod_insert_of_not_mem',
```
```suggestion
  rw finset.noncomm_prod_insert_of_not_mem' _ _ _ _ (not_mem_erase _ _),
```
mops up the extra goal you close with `simp` below

eric-wieser commented on src/data/finset/noncomm_prod.lean:
```diff
@@ -244,4 +280,39 @@ begin
+  ext x,
+  rw ← finset.noncomm_prod_single x,
+  rw finset.univ.noncomm_prod_map _ _ f,
+  rw finset.univ.noncomm_prod_map _ _ g,
```
You're in the `finset` namespace, so you don't need the prefix here (and elsewhere). Usually it's best to combine multiple `rw`s into a single `rw` too.

eric-wieser commented on src/data/finset/noncomm_prod.lean:
```diff
@@ -244,4 +280,39 @@ begin
+  rw ← finset.noncomm_prod_single x,
+  rw finset.univ.noncomm_prod_map _ _ f,
+  rw finset.univ.noncomm_prod_map _ _ g,
+  congr' 1, ext i, exact h i (x i),
```
```suggestion
  congr' 1 with i,
  exact h i (x i),
```

nomeata committed:
Apply code review

nomeata committed:
Apply more code review

nomeata committed:
Phrase lemmas wrt. pi.mul_single

nomeata committed:
Fix apply-lemmas for monoid_hom.single

nomeata committed:
Fix fallout

nomeata committed:
Throw in some to_additive attributes

nomeata commented on src/group_theory/subgroup/basic.lean:
```diff
@@ -1203,19 +1204,23 @@ begin
-lemma pi_mem_of_single_mem [fintype η] [decidable_eq η] {H : subgroup (Π i, f i) } (x : Π i, f i)
+@[to_additive]
+lemma pi_mem_of_mul_single_mem [fintype η] [decidable_eq η] {H : subgroup (Π i, f i) } (x : Π i, f i)
   (h : ∀ i, pi.mul_single i (x i) ∈ H) : x ∈ H :=
```
```suggestion
lemma pi_mem_of_mul_single_mem [fintype η] [decidable_eq η] {H : subgroup (Π i, f i)}
  (x : Π i, f i) (h : ∀ i, pi.mul_single i (x i) ∈ H) : x ∈ H :=
```

nomeata committed:
lint

nomeata committed:
lint

nomeata committed:
feat(data/list/prod_monoid): add prod_eq_pow_of_forall_eq

and rename `prod_le_of_forall_le` to `prod_le_pow_of_forall_le`

nomeata committed:
fix(src/algebra/group/pi): Fix apply-simp-lemmas for monoid_hom.single

so that the simp-normal form really is `pi.mul_single`.

While adjusting related lemmas in `group_theory.subgroup.basic`, add a
few missing `to_additive` attributes.

nomeata committed:
Merge branch 'joachim/single-simp-normal' into joachim/finite_pi_ext

nomeata committed:
Merge branch 'joachim/prod_eq_pow_of_forall_eq' into joachim/finite_pi_ext

nomeata committed:
Use npow, not pow, in name, so that to_additive does the right thing

nomeata committed:
Rename le_prod_of_forall_le to npow_le_prod_of_forall_le

nomeata committed:
Use `∀ (x ∈ l),` syntax

nomeata committed:
Try to teach to_additive about `pow` naming convention

nomeata committed:
Explicit name to avoid name clashes

nomeata committed:
More names are correct now

nomeata committed:
Even more names are correct now

nomeata committed:
Even more names are correct now (order_of_element)

nomeata committed:
Even more names are correct now (exponent.lean)

nomeata committed:
Even more names are correct now (topology/continuous_function/algebra.lean)

nomeata committed:
Say pow, not npow

nomeata committed:
Merge commit '28c902d86d7904b662d6caf6c9b211b05c485728' into joachim/single-simp-normal

nomeata committed:
Merge branch 'joachim/single-simp-normal' into joachim/finite_pi_ext

nomeata committed:
Fix sum_divisors_prime_pow

nomeata committed:
More correct names

nomeata committed:
Merge branch 'joachim/to_additive_pow' into joachim/prod_eq_pow_of_forall_eq

nomeata committed:
Merge commit 'eb46e7e3cf98edd45bd1cce2227edb90d7e4426d' into joachim/to_additive_pow

nomeata committed:
Merge branch 'joachim/to_additive_pow' into joachim/prod_eq_pow_of_forall_eq

nomeata committed:
Merge branch 'joachim/prod_eq_pow_of_forall_eq' into joachim/finite_pi_ext

nomeata committed:
Merge remote-tracking branch 'origin/master' into joachim/prod_eq_pow_of_forall_eq

nomeata committed:
Undo renaming, now happening in #12589

nomeata committed:
Rename the new lemma

nomeata committed:
Merge commit 'e3ad468bb46b2773596ec7329c4a298d71a1df08' into joachim/prod_eq_pow_of_forall_eq

nomeata committed:
Merge branch 'joachim/prod_eq_pow_of_forall_eq' into joachim/finite_pi_ext

nomeata committed:
Merge branch 'master' of github.com:leanprover-community/mathlib into joachim/finite_pi_ext

nomeata committed:
Rename lemmas according to latest fashion

nomeata committed:
Multisetify noncomm_prod_map

nomeata committed:
Multisetify noncomm_prod_eq_pow_card

nomeata committed:
Golf

nomeata committed:
Avoid rather specialized lemma

nomeata commented on src/data/finset/noncomm_prod.lean:
```diff
@@ -318,4 +365,38 @@ begin
+@[to_additive]
+lemma noncomm_prod_mul_single :
+  univ.noncomm_prod (λ i, pi.mul_single i (x i))
+  (λ i _ j _, by { by_cases h : i = j, { rw h }, { apply pi.mul_single_commute i j h } })
```
TODO: When are the laptop, check if this is just `pi.mul_single_apply_commute` now

nomeata committed:
Use pi.mul_single_apply_commute

jcommelin commented on src/algebra/group/pi.lean:
```diff
@@ -243,6 +244,31 @@ lemma pi.single_mul [Π i, mul_zero_class $ f i] (i : I) (x y : f i) :
+/-- The injection into a pi group at different indices commutes.
+
+For injections of commuting elements at the same index, see `commute.map` -/
+@[to_additive "The injection into an pi additive group at different indices commutes.
```
```suggestion
@[to_additive "The injection into an additive pi group at different indices commutes.
```

jcommelin reviewed (commented):
bors d+

nomeata commented on src/algebra/group/pi.lean:
```diff
@@ -243,6 +244,31 @@ lemma pi.single_mul [Π i, mul_zero_class $ f i] (i : I) (x y : f i) :
+end
+
+/-- The injection into a pi group with the same values commutes. -/
+@[to_additive "The injection into a pi additive group with the same values commutes."]
```
```suggestion
@[to_additive "The injection into an additive pi group with the same values commutes."]
```

nomeata committed:
Apply suggestions from code review

Co-authored-by: Johan Commelin <johan@commelin.net>
